File: lib/year2023/day03.rb
require "solver"
require "grid"
require "point"

module Year2023
  class Day03 < Solver
    def solve(part:)
      case part
      when 1 then schematic.parts.map(&:value).sum
      when 2 then schematic.gear_ratios.sum
      else nil
      end
    end

    def schematic
      @schematic = Schematic.new(lines)
    end

    class Schematic
      def initialize(lines)
        @lines = lines
      end

      def symbols
        @symbols ||= @lines.each_with_index.reduce(Hash.new) do |hash, (line, row)|
          line.chars.each_with_index do |ch, col|
            if ch != "." && !ch.match(/\d/)
              hash[[row, col]] = Symbol.new(row, col, ch)
            end
          end
          hash
        end
      end

      def numbers
        @numbers ||= @lines.size.times.reduce([]) do |array, row|
          col = 0
          while col < line(row).size
            number = Number.build(row, col, self)
            break if number.nil?
            number.set_symbol(self)
            array << number
            col = number.col + number.length
          end
          array
        end
      end

      def line(row)
        return "" if row < 0
        @lines[row] || ""
      end

      def substring(row, start, stop=-1)
        line(row)[start..stop] || ""
      end

      def char(row, col)
        return "." if col < 0
        line(row)[col] || "."
      end

      def parts
        numbers.select(&:part?)
      end

      def gear_parts
        @gear_parts ||= numbers.to_a.select(&:gear_part?)
      end

      def gear_ratios
        gear_parts.combination(2).map do |part, other|
          part.symbol == other.symbol ? part.value * other.value : 0
        end
      end

      class Number < Struct.new(:row, :col, :value, :length)
        attr_reader :symbol

        def self.build(row, col, schematic)
          m = schematic.substring(row, col).match(/\d+/)
          return if m.nil?
          value = m[0]
          col += schematic.substring(row, col).index(value)
          new(row, col, value.to_i, value.length)
        end

        def set_symbol(schematic)
          [
            [col - 1, row],
            [col + length, row],
            *(col-1..col+length).map { |c| [c, row - 1] },
            *(col-1..col+length).map { |c| [c, row + 1] },
          ].each do |(col, row)|
            schematic.symbols[[row, col]].tap do |symbol|
              @symbol = symbol unless symbol.nil?
            end
          end
        end

        def part?
          !@symbol.nil?
        end

        def gear_part?
          part? && @symbol.value == "*"
        end
      end

      class Symbol < Struct.new(:row, :col, :value)
      end
    end
  end
end

File: rust/src/year2023/day03.rs
use regex::Regex;
use std::collections::{HashMap, HashSet};

use crate::io::{Input, Solution};

use itertools::Itertools;

pub fn solve(part: &str, input: &Input) -> Solution {
    Solution::build(part, input, &part_one, &part_two)
}

type SchematicInput = Input;

impl SchematicInput {
    fn schematic(&self) -> Schematic {
        let symbols = self.symbols();
        let parts = self.parts(&symbols);
        Schematic { symbols, parts }
    }

    fn symbols(&self) -> HashMap<Loc, char> {
        let mut hashmap = HashMap::new();
        self.data.lines().enumerate().for_each(|(row, line)| {
            line.chars().enumerate().for_each(|(col, ch)| {
                if ch != '.' && !ch.is_ascii_digit() {
                    let loc = Loc { row, col };
                    hashmap.insert(loc, ch);
                }
            })
        });
        hashmap
    }

    fn parts(&self, symbols: &HashMap<Loc, char>) -> HashSet<Part> {
        let mut set = HashSet::new();
        let re = Regex::new(r"\d+").unwrap();
        self.data.lines().enumerate().for_each(|(row, line)| {
            for m in re.find_iter(line) {
                let value = m.as_str().parse::<u32>().unwrap();
                let mut neighbors = vec![(row, m.end())];
                if m.start() > 0 {
                    neighbors.push((row, m.start() - 1));
                }
                let start = if m.start() == 0 { 0 } else { m.start() - 1 };
                for c in start..=m.end() {
                    if row > 0 {
                        neighbors.push((row - 1, c));
                    }
                    neighbors.push((row + 1, c));
                }
                for (row, col) in neighbors {
                    let symbol_loc = Loc { row, col };
                    if symbols.contains_key(&symbol_loc) {
                        set.insert(Part { value, symbol_loc });
                        break;
                    }
                }
            }
        });
        set
    }
}

fn part_one(input: &Input) -> u32 {
    input
        .schematic()
        .parts
        .iter()
        .map(|part| part.value)
        .sum::<u32>()
}

fn part_two(input: &Input) -> u32 {
    input.schematic().gear_ratios().iter().sum::<u32>()
}

#[derive(Debug, Eq, PartialEq, Hash)]
struct Loc {
    row: usize,
    col: usize,
}

#[derive(Debug, Eq, PartialEq, Hash)]
struct Part {
    value: u32,
    symbol_loc: Loc,
}

#[derive(Debug)]
struct Schematic {
    symbols: HashMap<Loc, char>,
    parts: HashSet<Part>,
}

impl Schematic {
    fn gear_ratios(&self) -> Vec<u32> {
        self.parts
            .iter()
            .filter(|part| {
                self.symbols.contains_key(&part.symbol_loc)
                    && self.symbols.get(&part.symbol_loc).unwrap() == &'*'
            })
            .collect::<Vec<_>>()
            .iter()
            .combinations(2)
            .filter(|parts_| parts_[0].symbol_loc == parts_[1].symbol_loc)
            .map(|parts_| parts_[0].value * parts_[1].value)
            .collect()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn input() -> Input {
        Input::from(
            "\
            467..114..\n\
            ...*......\n\
            ..35..633.\n\
            ......#...\n\
            617*......\n\
            .....+.58.\n\
            ..592.....\n\
            ......755.\n\
            ...$.*....\n\
            .664.598..",
        )
    }

    #[test]
    fn test_part_one() {
        assert_eq!(part_one(&input()), 4361);
    }

    #[test]
    fn test_part_two() {
        assert_eq!(part_two(&input()), 467835);
    }
}
